wallet/src/Ergvein/Wallet/Monad/Front.hs

```diff
@@ -231,3 +235,52 @@ setFiltersSync :: MonadFrontAuth t m => Currency -> Bool -> m ()
 setFiltersSync c v = do
   r <- getFiltersSyncRef
   modifyExternalRef r $ (, ()) . M.insert c v
+
+requestRandomIndexer :: MonadFront t m => Event t (Currency, Message) -> m (Event t (SockAddr, Message))
+requestRandomIndexer reqE = mdo
+  let actE = leftmost [Just <$> reqE, Nothing <$ sentE]
+  sentE <- widgetHoldE (pure never) $ ffor actE $ \case
+    Nothing -> pure never
+    Just (cur, req) -> requester cur req
+  pure sentE
+
+requester :: MonadFront t m => Currency -> Message -> m (Event t (SockAddr, Message))
+requester cur req = mdo
+  connsD  <- externalRefDynamic =<< getActiveConnsRef
+  buildE <- getPostBuild
+  respD <- holdDyn True $ False <$ respE
+  te <- widgetHoldDynE $ ffor respD $ \b -> if b
+    then tickLossyFromPostBuildTime timeout
+    else pure never
+  let goE = leftmost [buildE, () <$ te]
+  respE <- widgetHoldE (pure never) $ ffor goE $ const $ do
+    mconn <- randomElem =<< getOpenSyncedConns cur
+    case mconn of
+      Nothing -> pure never
+      Just conn -> do
+        requestIndexerWhenOpen conn req
+        pure $ (indexConAddr conn,) <$> indexConRespE conn
+  pure respE
+  where
+    timeout = 5 -- NominalDiffTime, seconds
+
+-- | Designed to be used inside a widgetHold, samples dynamics
+getOpenSyncedConns :: MonadFront t m => Currency -> m [IndexerConnection t]
+getOpenSyncedConns cur = do
+  conns <- readExternalRef =<< getActiveConnsRef
+  heights <- readExternalRef =<< getHeightRef
+  let mh = M.lookup cur heights
+  fmap catMaybes $ flip traverse (M.elems conns) $ \con -> do
+    isUp <- sampleDyn $ indexConIsUp con
+    if not isUp then pure Nothing else do
+      mh' <- fmap (M.lookup cur) $ sampleDyn $ indexerConHeight con
+      pure $ case (mh, fromIntegral <$> mh') of
+        (Just h, Just h') -> if h == h' || h - h' == 1 then Just con else Nothing
+        _ -> Nothing
+
+randomElem :: MonadIO m => [a] -> m (Maybe a)
```

Comment:
Duplication of https://github.com/hexresearch/ergvein/blob/d30370786dc0f83b4566133acb313380a9bfe505/wallet/src/Ergvein/Wallet/Monad/Client.hs#L214 ?

Reply:
No. randomElem was moved from Ergvein.Wallet.Monad.Client to Ergvein.Wallet.Monad.Front

requestRandomIndexer now requires MonadFront, not MonadIndexClient